api/example.com/resource/gpu/v1alpha1/api.go

```diff
@@ -17,32 +17,91 @@
 package v1alpha1
 
 import (
-	"k8s.io/utils/ptr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
 )
 
 const (
 	GroupName = "gpu.resource.example.com"
 	Version   = "v1alpha1"
 
-	GpuDeviceType     = "gpu"
-	UnknownDeviceType = "unknown"
-
-	GpuClaimParametersKind = "GpuClaimParameters"
+	GpuConfigKind = "GpuConfig"
 )
 
-func DefaultDeviceClassParametersSpec() *DeviceClassParametersSpec {
-	return &DeviceClassParametersSpec{
-		DeviceSelector: []DeviceSelector{
-			{
-				Type: GpuDeviceType,
-				Name: "*",
+// Decoder implements a decoder for objects in this API group.
+var Decoder runtime.Decoder
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// GpuConfig holds the set of parameters for configuring a GPU.
+type GpuConfig struct {
+	metav1.TypeMeta `json:",inline"`
+	Sharing         *GpuSharing `json:"sharing,omitempty"`
+}
+
+// DefaultGpuConfig provides the default GPU configuration.
+func DefaultGpuConfig() GpuConfig {
+	return GpuConfig{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: GroupName + "/" + Version,
+			Kind:       GpuConfigKind,
+		},
+		Sharing: &GpuSharing{
+			Strategy: TimeSlicingStrategy,
+			TimeSlicingConfig: &TimeSlicingConfig{
+				Duration: "Default",
 			},
 		},
 	}
 }
 
-func DefaultGpuClaimParametersSpec() *GpuClaimParametersSpec {
-	return &GpuClaimParametersSpec{
-		Count: ptr.To(1),
+// Normalize updates a GpuConfig config with implied default values based on other settings.
+func (c *GpuConfig) Normalize() {
+	if c == nil {
+		*c = DefaultGpuConfig()
```

Comment:
`c` is nil, so this will panic.

Reply:
Or simply document it as "mutates the instance and thus must not be called with nil"?

Reply:
Then calling it for nil will panic because its a programming mistake.

Reply:
I just added an error -- it makes it more symmetrical with other code around it anyway.